kaiax/valset/impl/getter_proposers.go

```diff
@@ -0,0 +1,237 @@
+	}
+	logger.Debug("GetProposerList", "number", c.num, "list", valset.NewAddressSet(list).String())
+
+	// Store to cache
+	v.proposerListCache.Add(updateNum, list)
+	return list, nil
+}
+
+func (v *ValsetModule) getRemoveVotesInInterval(pUpdateNum, pUpdateInterval uint64) removeVoteList {
+	// return early if removeVoteLists is found in cache
+	if list, ok := v.removeVotesCache.Get(pUpdateNum); ok {
+		return list.(removeVoteList)
+	}
+
+	// collect remove validator votes by scanning blocks between [pUpdateNum, pUpdateNum+pUpdateInterval)
+	scanBlocks := v.scanBlocks(pUpdateNum, pUpdateInterval)
+	removeVotes := make(removeVoteList, 0)
+	for _, num := range scanBlocks {
+		header := v.Chain.GetHeaderByNumber(num)
+		if header == nil || len(header.Vote) == 0 {
+			continue // skip nil header or vote
+		}
+		voteKey, addresses, ok := parseValidatorVote(header)
+		if !ok {
+			logger.Error("Failed to parse validator vote", "block", num)
+			return nil
+		}
+		if len(addresses) > 0 && voteKey == gov.RemoveValidator {
+			// demoted validators are not considered
+			qualified, err := v.getQualifiedValidators(num)
+			if err != nil {
+				logger.Error("Failed to get qualified validators", "block", num, "err", err)
+				return nil
+			}
+			removeVotes.add(num, addresses, qualified)
+		}
+	}
+
+	// Store to cache
+	if len(removeVotes) > 0 {
+		v.removeVotesCache.Add(pUpdateNum, removeVotes)
+	}
+	return removeVotes
+}
+
+// scanBlocks returns the block numbers to be scanned for remove validator votes
+func (v *ValsetModule) scanBlocks(pUpdateNum, pUpdateInterval uint64) []uint64 {
+	scanBlocks := make([]uint64, 0)
+
+	// if migrated, scanBlocks is set as voteBlockNums between [pUpdateNum, pUpdateNum+pUpdateInterval)
+	pMinVoteNum := ReadLowestScannedVoteNum(v.ChainKv)
+
+	if pMinVoteNum != nil && *pMinVoteNum <= pUpdateNum {
+		if v.validatorVoteBlockNumsCache != nil {
+			scanBlocks = make([]uint64, len(v.validatorVoteBlockNumsCache))
+			copy(scanBlocks, v.validatorVoteBlockNumsCache)
+		} else {
+			scanBlocks = ReadValidatorVoteBlockNums(v.ChainKv)
+		}
+		scanBlocks = slices.DeleteFunc(scanBlocks, func(n uint64) bool {
+			return !(n >= pUpdateNum && n < pUpdateNum+pUpdateInterval)
+		})
+		return scanBlocks
+	}
+
+	// if not migrated, scanBlocks is set as blocknums between [updateNum, pUpdateNum+pUpdateInterval)
+	for i := pUpdateNum; i < pUpdateNum+pUpdateInterval; i++ {
+		scanBlocks = append(scanBlocks, i)
+	}
+	return scanBlocks
+}
+
+type removeVote struct {
+	voteBlkNum  uint64
+	removeAddrs []common.Address
+}
+type removeVoteList []removeVote
+
+func (r *removeVoteList) add(voteBlkNum uint64, addresses []common.Address, qualified *valset.AddressSet) {
+	removeAddrs := valset.NewAddressSet([]common.Address{})
+	for _, addr := range addresses {
+		if qualified.Contains(addr) {
+			removeAddrs.Add(addr)
+		}
+	}
+	if removeAddrs.Len() > 0 {
+		*r = append(*r, removeVote{voteBlkNum, removeAddrs.List()})
+	}
```

Comment:
Can you add a sort function call here? It seems `removeVoteList` must always be sorted

Reply:
Or you can just add a comment for now saying it must be called sequentially

Reply:
Oh right, instead of sorting it, i've added sort checks at `add`, `filteredProposerList` methods of `removeVoteList`. https://github.com/kaiachain/kaia/pull/183/commits/2ef078d4ef7b23c0300146b85a24f0c5515ded4e
It will guarantee the sorting of the `removevoteList`.